### PR TITLE
Events

### LINKED ISSUE #122: Events
- [ ] track of latest topic addressed
- [ ] keep historical of status
- [ ] trigger of a push on event to a processor

### PR DESCRIPTION
**Pull request by maasencioh**

### CODE REVIEW

File: addons/events/Queries/nextEvent.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+): Promise<Event> {
+  return Event.query(
+    {
+      topic,
+      processors: {
+        $elemMatch: {
+          processorId,
+          history: {
+            $or: [
+              { $elemMatch: { status: EventStatus.PENDING } },
```
Comment (targos) [inline addons/events/Queries/nextEvent.ts]: It's only the first element that should be pending here

File: addons/events/Models/Event.ts [hunk 2]
```diff
@@ -26,7 +26,9 @@ interface EventData {
   fileId: string;
 }
 
-export default class Event extends BaseModel {
+export class Event extends BaseModel {
+  public static collectionName = 'events';
```
Comment (targos) [inline addons/events/Models/Event.ts]: why is this needed?

Comment (maasencioh) [reply addons/events/Models/Event.ts]: I don't know, but without it no collection was created

Comment (targos) [reply addons/events/Models/Event.ts]: the collection name is supposed to be computed automatically if not set. Please investigate this.

Comment (maasencioh) [reply addons/events/Models/Event.ts]: seems to be working now, I think that was related with another step

Comment (targos) [reply addons/events/Models/Event.ts]: okay. when you have issues with our modules, don't hesitate to talk to us instead of working around them

File: addons/events/routes.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import setEventStatus from './Queries/setEventStatus';
+
+Route.get(
+  '/next-event',
+  async ({ request, response, auth }: HttpContextContract) => {
+    // Validates that is already authenticated
+    const isAuth = await auth.check();
+    if (!isAuth) {
+      return response.internalServerError({ errors: ['Not authorized'] });
+    }
```
Comment (targos) [inline addons/events/routes.ts]: I think there is another method that throws if user is not authenticated

File: addons/events/routes.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    // Checks that the params are valid
+    const params: Record<string, string | undefined> = request.all();
+    if (!params.processorId) {
+      return response.internalServerError({
+        errors: ['Processor not specified'],
+      });
+    }
+    if (!params.topic) {
+      return response.internalServerError({ errors: ['Topic not specified'] });
+    }
```
Comment (targos) [inline addons/events/routes.ts]: use a validator: https://docs.adonisjs.com/guides/validator/introduction#validator-classes

File: addons/events/routes.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    const event = await nextEvent(params.processorId, params.topic);
+    return response.json(event);
+  },
+);
+
+Route.put(
+  '/set-event',
+  async ({ request, response, auth }: HttpContextContract) => {
+    // Validates that is already authenticated
+    const isAuth = await auth.check();
```
Comment (targos) [inline addons/events/routes.ts]: ditto

File: addons/events/routes.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      return response.internalServerError({ errors: ['Event not specified'] });
+    }
+    if (!params.processorId) {
+      return response.internalServerError({
+        errors: ['Processor not specified'],
+      });
+    }
+    if (!params.status) {
+      return response.internalServerError({ errors: ['Status not specified'] });
+    }
```
Comment (targos) [inline addons/events/routes.ts]: ditto

File: addons/events/Queries/nextEvent.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+          processorId,
+          'history.0': {
+            $or: [{ status: EventStatus.PENDING }, { $exist: false }],
+          },
+        },
+      },
+    },
+    {
+      driverOptions: { sort: { _id: -1 } },
+    },
```
Comment (targos) [inline addons/events/Queries/nextEvent.ts]:
```suggestion
```

That's the default sort order for queries

**Review by targos — APPROVED** [review]
